src/strings/nfa-noodlification.cc

```diff
@@ -445,11 +492,25 @@ std::vector<seg_nfa::TransducerNoodle> seg_nfa::noodlify_for_transducer(
     // of the transducer, so that intersection works correctly (i.e. the delimiters behave
     // as epsilon transitions).
 
-    Nft intersection = *nft;
-
-    // we intersect input nfa with nft on the input track but we need to add INPUT_DELIMITER as an "epsilon transition" of nft
-    add_self_loop_for_every_default_state(intersection, INPUT_DELIMITER);
-    intersection = mata::nft::compose(concatenated_input_nft, intersection, 0, 0, false);
+    Nft intersection;
+
+    if (is_nft_homomorphic(nft)) {
+        // For the case that input nft T is homomorphic, i.e., for each two words u,v, we have T(u.v) = T(u).T(v),
+        // we can do the following optimization. Let I1, ..., In be the input automata. Because T is homomorphic,
+        // we can take the concatenation T(I1).T(I2)...T(In) connected with INPUT_DELIMITER instead of computing
+        // the composition with concatenated_input_nft.
+        mata::nft::Nft concatenation{mata::nft::compose(mata::nft::Nft(*input_automata[0]), *nft, 0, 0, false)};
+        for (std::vector<std::shared_ptr<mata::nfa::Nfa>>::size_type i = 1; i < input_automata.size(); ++i) {
```

Comment:
Just `size_t` instead of the `size_type`, I would say. For better readability. Or `auto`, if you want to be exact.